File: file/310.js
 ============================================================================
 *   常量
 * ============================================================================ */
//const HOST = 'https://www.lingshid.net';
const HOST = 'https://www.cnrbc.com';
const LOGO = 'https://www.cnrbc.com/favicon.ico';
const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36';

/* ============================================================================
 *   工具函数
 * ============================================================================ */
function clean(s) {
    return String(s || '').replace(/<script[\s\S]*?<\/script>/gi, '')
        .replace(/<style[\s\S]*?<\/style>/gi, '')
        .replace(/<[^>]+>/g, '')
        .replace(/&nbsp;/g, ' ').replace(/&amp;/g, '&')
        .replace(/&quot;/g, '"').replace(/&#39;/g, "'")
        .replace(/\s+/g, ' ').trim();
}

function absUrl(url) {
    if (!url) return LOGO;
    url = String(url).trim();
    if (/^https?:\/\//i.test(url)) return url;
    if (url.indexOf('//') === 0) return 'https:' + url;
    if (url[0] === '/') return HOST + url;
    return HOST + '/' + url;
}

function safeJson(t, d) {
    try {
        return JSON.parse(t || '{}');
    } catch (e) {
        return d || {};
    }
}

function getHeaders() {
    return {
        'User-Agent': UA,
        'Referer': HOST + '/',
        'Accept': '*/*',
        'Accept-Language': 'zh-CN,zh;q=0.9'
    };
}

/* ── 网络请求 ── */
async function fetchText(url, timeout) {
    const hd = getHeaders();
    const to = timeout || 10000;
    try {
        if (typeof Java !== 'undefined' && Java && Java.req) {
            const r = await Java.req(url, {
                headers: hd,
                timeout: to
            });
            if (typeof r === 'string') return r;
            return String((r && (r.body || r.content || r.data)) || '');
        }
        const r2 = await req(url, {
            headers: hd,
            timeout: to
        });
        if (typeof r2 === 'string') return r2;
        return String((r2 && (r2.content || r2.body || r2.data)) || '');
    } catch (e) {
        return '';
    }
}

async function fetchHtml(url) {
    return await fetchText(url, 10000);
}

/* ── 从HTML解析单场比赛 ── */
function parseOneMatch(block) {
    if (!block) return null;

    // 1. 提取 href（完整URL）和 title
    const hrefM = /href="(https?:\/\/[^"]*\/detail\/\d+)"/.exec(block);
    const titleM = /title="([^"]*)"/.exec(block);
    if (!hrefM) return null;

    const vodId = hrefM[1];
    const rawTitle = titleM ? clean(titleM[1]) : '';

    // 去掉末尾日期 (2026-06-26 ) → 只保留"主队 vs 客队"
    let displayName = rawTitle.replace(/\s*\(\d{4}-\d{1,2}-\d{1,2}\s*\)\s*$/, '').trim();

    // 2. 提取 data-time 属性（日期）
    let dateStr = '';
    const dateM = /data-time\s*=\s*"([^"]*)"/.exec(block);
    if (dateM) dateStr = dateM[1].trim();

    // 3. 提取主队Logo —— 找 team zhudui 中的 img src
    let pic = LOGO;
    const zhuduiM = /<div\s+class="team\s+zhudui"[^>]*>([\s\S]*?)<\/div>\s*<div\s+class="center"/.exec(block);
    if (zhuduiM) {
        const zhuduiHtml = zhuduiM[1];
        const imgM = /<img[^>]*src="([^"]+)"[^>]*>/.exec(zhuduiHtml);
        if (imgM) pic = absUrl(imgM[1]);
    }

    // 4. 提取联赛名和时间
    let league = '';
    let time = '';
    const infoM = /<em>([^<]*)<\/em>\s*<i>([^<]*)<\/i>/.exec(block);
    if (infoM) {
        league = clean(infoM[1]);
        time = clean(infoM[2]);
    }

    // 5. 判断是否正在直播中 —— 仅检测 zb_green 类
    let isLive = false;
    if (block.includes('zb_green')) isLive = true;

    // 备注 = 🟢 + 联赛名 + 日期 + 时间
    let remark = '';
    if (isLive) remark = '🟢 ';
    if (league) remark += league;
    if (dateStr) remark += ' ' + dateStr;
    if (time) remark += ' ' + time;
    remark = remark.trim();

    // 运动类型判断（通过图标sport_1=足球 sport_2=篮球）
    let sportType = 'football';
    if (block.includes('sport_2.png')) sportType = 'basketball';

    return {
        vod_id: vodId,
        vod_name: displayName || rawTitle,
        vod_pic: pic,
        vod_remarks: remark,
        vod_content: (league || '体育直播') + '\n' + displayName,
        type_name: league || '全部',
        _sport: sportType,
        _league: league
    };
}

/* ── 从HTML解析全部比赛 ── */
function parseMatches(html) {
    const matches = [];
    if (!html) return matches;

    const cardRegex = /<a\s+class="clearfix\s*"[^>]*href="(https?:\/\/[^"]*\/detail\/\d+)"[^>]*>([\s\S]*?)<\/a>/g;
    let m;
    while ((m = cardRegex.exec(html)) !== null) {
        const block = m[0];

        // 跳过非直播卡片
        if (!block.includes('video-icon') && !block.includes('视频直播')) continue;

        const match = parseOneMatch(block);
        if (match) matches.push(match);
    }

    return matches;
}

/* ── 从详情页提取播放地址 ── */
function extractPlayUrl(html) {
    if (!html) return '';

    // 方式1: data-play 属性
    let m = /data-play\s*=\s*["']([^"']+)["']/.exec(html);
    if (m) return m[1];

    // 方式2: playurl 变量
    m = /playurl\s*[:=]\s*["']([^"']+)["']/.exec(html);
    if (m) return m[1];

    // 方式3: iframe src 包含 .m3u8
    m = /<iframe[^>]+src=["']([^"']*\.m3u8[^"']*)["']/.exec(html);
    if (m) return m[1];

    // 方式4: video src 包含 .m3u8
    m = /<video[^>]+src=["']([^"']*\.m3u8[^"']*)["']/.exec(html);
    if (m) return m[1];

    // 方式5: 任意 .m3u8 链接
    m = /https?:\/\/[^"'\s]+\.m3u8[^"'\s]*/.exec(html);
    if (m) return m[0];

    // 方式6: vid 参数
    m = /vid\s*[:=]\s*["']([^"']+)["']/.exec(html);
    if (m) {
        const vid = m[1];
        if (vid.length > 5) return 'https://play.cnrbc.com/play/' + vid + '.m3u8';
    }

    return '';
}

/* ============================================================================
 *   分类定义
 * ============================================================================ */
function getClasses() {
    return [{
            type_id: 'all',
            type_name: '全部'
        },
        {
            type_id: 'zuqiu',
            type_name: '⚽ 足球'
        },
        {
            type_id: 'lanqiu',
            type_name: '🏀 篮球'
        }
    ];
}

function isFootball(league, sport) {
    if (sport === 'football') return true;
    if (sport === 'basketball') return false;
    const name = (league || '').toLowerCase();
    if (/篮球|nba|wnba|cba|nbl|basketball/.test(name)) return false;
    return true;
}

function isBasketball(league, sport) {
    if (sport === 'basketball') return true;
    const name = (league || '').toLowerCase();
    return /篮球|nba|wnba|cba|nbl|basketball/.test(name);
}

/* ============================================================================
 *   爬虫接口
 * ============================================================================ */
async function init(cfg) {}

async function home(filter) {
    return JSON.stringify({
        class: getClasses(),
        filters: {},
        style: {
            type: 'list',
            ratio: 1
        }
    });
}

async function homeVod() {
    return await category('all', 1, {}, {});
}

async function category(tid, pg, filter, extend) {
    tid = String((extend && extend.cateId) || tid || 'all');
    pg = parseInt(pg) || 1;
    let list = [];

    try {
        // 统一用首页解析
        const html = await fetchHtml(HOST);
        if (!html) {
            return JSON.stringify({
                code: 1,
                msg: 'ok',
                page: 1,
                pagecount: 1,
                limit: 30,
                total: 0,
                list: [],
                style: {
                    type: 'list',
                    ratio: 1
                }
            });
        }

        let matches = parseMatches(html);

        // 用 _sport 字段进行过滤
        if (tid !== 'all') {
            matches = matches.filter(item => {
                const league = item._league || '';
                const sport = item._sport || '';
                if (tid === 'zuqiu') return isFootball(league, sport);
                if (tid === 'lanqiu') return isBasketball(league, sport);
                return true;
            });
        }

        list = matches;

    } catch (e) {
        // 静默
    }

    return JSON.stringify({
        code: 1,
        msg: '赛程列表',
        page: pg,
        pagecount: 1,
        limit: 200,
        total: list.length,
        list: list,
        style: {
            type: 'list',
            ratio: 1
        }
    });
}

async function detail(ids) {
    const id = String(Array.isArray(ids) ? ids[0] : ids || '');
    if (!id) {
        return JSON.stringify({
            code: 1,
            msg: 'ok',
            page: 1,
            pagecount: 1,
            limit: 1,
            total: 0,
            list: []
        });
    }

    try {
        const html = await fetchHtml(id);
        if (!html) {
            return JSON.stringify({
                code: 1,
                msg: 'ok',
                page: 1,
                pagecount: 1,
                limit: 1,
                total: 0,
                list: []
            });
        }

        // 提取名称：严格从详情页HTML中提取"主队 vs 客队"
        let name = '';

        // 从页面中找 h1 或 h2 标题标签
        const hM = /<h[12][^>]*>([\s\S]*?)<\/h[12]>/i.exec(html);
        if (hM) {
            let hText = clean(hM[1]);
            // 匹配 "主队 vs 客队"
            const vsM = /^(.+?)\s*(?:vs|VS)\s*(.+?)$/.exec(hText);
            if (vsM) {
                name = vsM[1].trim() + ' vs ' + vsM[2].trim();
            }
        }

        // fallback: 从 title 中提取
        if (!name) {
            const titleM = /<title>([^<]+)<\/title>/.exec(html);
            if (titleM) {
                let t = clean(titleM[1]);
                // 严格匹配 "球队名 vs 球队名"（最多各3个词，且不含站点关键词）
                const strictVs = /^(?:[A-Za-z\u4e00-\u9fa5][A-Za-z\u4e00-\u9fa5\s()]*?)\s+(?:vs|VS)\s+(?:[A-Za-z\u4e00-\u9fa5][A-Za-z\u4e00-\u9fa5\s()]*?)$/;
                const m = strictVs.exec(t);
                if (m) {
                    name = m[0].trim();
                } else {
                    // 宽松匹配：找 "主队-vs-客队" 或 "主队 vs 客队"
                    const looseVs = /^(.+?)\s*[-–—]\s*(?:vs|VS)\s*[-–—]\s*(.+?)(?:\s*\([\d\-]+\))?\s*$/;
                    const m2 = looseVs.exec(t);
                    if (m2) {
                        name = m2[1].trim() + ' vs ' + m2[2].trim();
                    }
                }
            }
        }

        // 最后fallback
        if (!name) name = '直播';

        const playUrl = extractPlayUrl(html);

        let pic = LOGO;
        const picM = /<img[^>]*src="([^"]*)"[^>]*class="lazy"[^>]*>/.exec(html);
        if (picM) pic = absUrl(picM[1]);

        const vod = {
            vod_id: id,
            vod_name: name,
            vod_pic: pic,
            vod_remarks: '',
            vod_content: name || '体育直播',
            vod_play_from: '310直播',
            vod_play_url: playUrl ? '直播地址$' + playUrl : '暂无信号$' + HOST
        };

        return JSON.stringify({
            code: 1,
            msg: '直播详情',
            page: 1,
            pagecount: 1,
            limit: 1,
            total: 1,
            list: [vod]
        });
    } catch (e) {
        return JSON.stringify({
            code: 1,
            msg: 'ok',
            page: 1,
            pagecount: 1,
            limit: 1,
            total: 0,
            list: []
        });
    }
}

async function search(wd, quick, pg) {
    return JSON.stringify({
        code: 1,
        msg: '搜索',
        page: 1,
        pagecount: 1,
        limit: 20,
        total: 0,
        list: []
    });
}

async function play(flag, id, flags) {
    const url = String(id || '');
    if (/\.(m3u8|flv|mp4)(\?|$)/i.test(url)) {
        return JSON.stringify({
            parse: 0,
            url: url,
            header: {
                'User-Agent': UA,
                'Referer': HOST + '/',
                'Origin': HOST
            }
        });
    }
    return JSON.stringify({
        parse: 1,
        url: url,
        header: {
            'User-Agent': UA,
            'Referer': HOST + '/',
            'Origin': HOST
        }
    });
}

/* ============================================================================
 *   包装函数（统一响应格式）
 * ============================================================================ */
async function homeContent(f) {
    return safeJson(await home(f), {
        class: [],
        filters: {}
    });
}

async function homeVideoContent() {
    return safeJson(await homeVod(), {
        list: []
    });
}

async function categoryContent(tid, pg, f, ext) {
    return safeJson(await category(tid, pg, f, ext || {}), {
        list: []
    });
}

async function detailContent(ids) {
    return safeJson(await detail(ids), {
        list: []
    });
}

async function searchContent(wd, q, pg) {
    return safeJson(await search(wd, q, pg || 1), {
        list: []
    });
}

async function playerContent(flag, id, flags) {
    return safeJson(await play(flag, id, flags), {
        parse: 1,
        url: id
    });
}

/* ============================================================================
 *   引擎导出 — 全局函数模式
 * ============================================================================ */
export function __jsEvalReturn() {
    return {
        init: init,
        home: home,
        homeVod: homeVod,
        category: category,
        search: search,
        detail: detail,
        play: play,
        homeContent: homeContent,
        homeVideoContent: homeVideoContent,
        categoryContent: categoryContent,
        detailContent: detailContent,
        searchContent: searchContent,
        playerContent: playerContent
    };
}

File: file/310.py
#!/usr/bin/env python3
# 310直播 V1.1 - remarks格式: 🔴 MM-DD HH:MM league
import re, json, time, requests
from base.spider import Spider as BaseSpider

HOST = 'https://www.cnrbc.com'
LOGO = HOST + '/favicon.ico'
UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36'


class Spider(BaseSpider):
    def getName(self):
        return "310直播"

    def init(self, extend=""):
        self.host = HOST
        self._cache = {}
        self._CACHE_TTL = 120

    def _get_cache(self, key):
        e = self._cache.get(key)
        if e and time.time() - e['t'] < self._CACHE_TTL:
            return e['v']
        return None

    def _set_cache(self, key, val):
        self._cache[key] = {'v': val, 't': time.time()}

    def _headers(self):
        return {'User-Agent': UA, 'Referer': HOST + '/', 'Accept': 'text/html,application/xhtml+xml'}

    def _get_html(self, url):
        try:
            r = requests.get(url, headers=self._headers(), timeout=15)
            r.encoding = r.apparent_encoding or 'utf-8'
            return r.text
        except:
            return ''

    def _parse_cards(self, html):
        if not html: return []
        results = []
        for m in re.finditer(r'<a[^>]*class="clearfix[^"]*"[^>]*href="(https?://[^"]+)"[^>]*>(.*?)</a>', html, re.DOTALL):
            href, body = m.group(1), m.group(2)
            tag_text = m.group(0)[:m.group(0).find('>') + 1]
            if 'video-icon' not in body and '视频直播' not in body:
                continue
            mid_m = re.search(r'/detail/(\d+)', href)
            if not mid_m: continue
            mid = mid_m.group(1)
            home_m = re.search(r'team zhudui.*?<p>([^<]+)', body, re.DOTALL)
            away_m = re.search(r'team kedui.*?<p>([^<]+)', body, re.DOTALL)
            home = home_m.group(1).strip() if home_m else ''
            away = away_m.group(1).strip() if away_m else ''
            if not home or not away: continue
            league_m = re.search(r'<em>([^<]+)</em>\s*<i>([^<]+)', body)
            league = league_m.group(1).strip() if league_m else ''
            match_time = league_m.group(2).strip() if league_m else ''
            date_m = re.search(r'data-time="([^"]+)"', tag_text)
            data_date = date_m.group(1).strip() if date_m else ''
            logo_m = re.search(r'team zhudui.*?src="([^"]+)".*?alt="([^"]+)"', body, re.DOTALL)
            pic = ''
            if logo_m:
                src = logo_m.group(1)
                pic = src if src.startswith('http') else HOST + '/' + src.lstrip('/')
            is_live = 'zb_green' in body
            sport_m = re.search(r'sport_(\d)\.png', body)
            sport_type = 'basketball' if sport_m and sport_m.group(1) == '2' else 'football'
            tag_icon = '\U0001f534' if is_live else ''
            date_part = data_date[5:] if data_date else ''
            parts = []
            if tag_icon: parts.append(tag_icon)
            if date_part: parts.append(date_part)
            if match_time: parts.append(match_time)
            if league: parts.append(league)
            results.append({
                'mid': mid, 'home': home, 'away': away,
                'league': league, 'time': match_time, 'date': data_date, 'pic': pic,
                'is_live': is_live, 'sport_type': sport_type,
                'remarks': ' '.join(parts) or '直播',
            })
        return results

    def _get_classes(self):
        return [
            {'type_id': 'all', 'type_name': '全部'},
            {'type_id': 'zuqiu', 'type_name': '⚽ 足球'},
            {'type_id': 'lanqiu', 'type_name': '🏀 篮球'},
        ]

    def homeContent(self, filter):
        return {'class': self._get_classes(), 'filters': {}}

    def categoryContent(self, tid, pg, filter, extend):
        try:
            tid = (extend or {}).get('cateId') or tid or 'all'
            pg = int(pg or 1)
            cached = self._get_cache(f'cards_{tid}')
            if cached is not None:
                cards = cached
            else:
                html = self._get_html(HOST)
                all_cards = self._parse_cards(html)
                self._set_cache('cards_all', all_cards)
                cards = all_cards
                if tid == 'zuqiu':
                    cards = [c for c in all_cards if c['sport_type'] == 'football']
                elif tid == 'lanqiu':
                    cards = [c for c in all_cards if c['sport_type'] == 'basketball']
                else:
                    cards = all_cards
                self._set_cache(f'cards_{tid}', cards)
            list_items = []
            for c in cards:
                list_items.append({
                    'vod_id': f'310|{c["mid"]}',
                    'vod_name': f'{c["home"]} vs {c["away"]}',
                    'vod_pic': c['pic'] or LOGO,
                    'vod_remarks': c['remarks'],
                    'type_name': c['league'] or '全部',
                })
            return {
                'page': pg, 'pagecount': 1, 'limit': len(list_items),
                'total': len(list_items), 'list': list_items,
            }
        except:
            return {'page': 1, 'pagecount': 1, 'limit': 30, 'total': 0, 'list': []}

    def detailContent(self, ids):
        try:
            vid = (ids[0] if isinstance(ids, list) else str(ids or '')).strip()
            if not vid: return {'list': []}
            parts = vid.split('|')
            mid = parts[1] if len(parts) > 1 else parts[0]
            url = f'{HOST}/detail/{mid}'
            html = self._get_html(url)
            if not html: return {'list': []}
            name = ''
            h1_m = re.search(r'<h[12][^>]*>([\s\S]*?)</h[12]>', html)
            if h1_m:
                h_text = re.sub(r'<[^>]+>', '', h1_m.group(1)).strip()
                h_text = re.sub(r'^正在直播：\s*', '', h_text)
                vs_m = re.match(r'^(?:\d{2}-\d{2}\s+\d{2}:\d{2}\s+)?(.+?)\s*[-–—]?\s*vs\s*[-–—]?\s*(.+?)(?:\s*[\(（].*?[\)）])?\s*$', h_text)
                if vs_m:
                    name = vs_m.group(1).strip() + ' vs ' + vs_m.group(2).strip()
            if not name:
                title_m = re.search(r'<title>([^<]+)', html)
                if title_m:
                    t = title_m.group(1).strip()
                    t = re.sub(r'【310直播】.*?为您在线直播', '', t).strip()
                    t = re.sub(r'-310直播$', '', t).strip()
                    vs_m3 = re.match(r'^(.+?)\s*vs\s*(.+?)$', t)
                    if vs_m3:
                        name = vs_m3.group(1).strip() + ' vs ' + vs_m3.group(2).strip()
                    else:
                        name = t
            if not name: name = '直播'
            play_url = ''
            m3u8_m = re.search(r"src:\s*'([^']*\.m3u8[^']*)'", html)
            if m3u8_m:
                play_url = m3u8_m.group(1).replace('&amp;', '&')
            if not play_url:
                m3u8_m2 = re.search(r"https?://[^\"'\s]+\.m3u8[^\"'\s]*", html)
                if m3u8_m2: play_url = m3u8_m2.group(0)
            pic = LOGO
            img_m = re.search(r'<img[^>]*src="([^"]*)"[^>]*class="lazy"', html)
            if not img_m:
                img_m = re.search(r'team\s+zhudui.*?<img[^>]*src="([^"]+)"', html, re.DOTALL)
            if not img_m:
                img_m = re.search(r'<img[^>]*src="([^"]*badge/competitors[^"]+)"', html)
            if img_m:
                src = img_m.group(1)
                pic = src if src.startswith('http') else HOST + '/' + src.lstrip('/')
            if play_url == '': play_url = HOST
            has_signal = play_url != HOST
            return {'list': [{
                'vod_id': vid,
                'vod_name': name,
                'vod_pic': pic,
                'vod_remarks': '',
                'vod_play_from': '310直播',
                'vod_play_url': f'直播地址${play_url}' if has_signal else f'暂无信号${HOST}',
                'vod_content': name,
            }]}
        except:
            return {'list': []}

    def searchContent(self, kw, quick, pg):
        return {'page': 1, 'pagecount': 1, 'limit': 20, 'total': 0, 'list': []}

    def playerContent(self, flag, id, vipFlags):
        try:
            url = str(id or '')
            parse = 0 if re.search(r'\.(m3u8|flv|mp4)(\?|$)', url, re.IGNORECASE) else 1
            return {'parse': parse, 'url': url, 'header': {
                'User-Agent': UA, 'Referer': HOST + '/', 'Origin': HOST
            }}
        except:
            return {'parse': 1, 'url': str(id or '')}

    def localProxy(self, param):
        return [200, "text/plain", b"", {}]
